### PR TITLE
Initial commit of CNCF conformance test scripts

### PR DESCRIPTION
**Pull request by rolpat**

### CODE REVIEW

File: .ci/run-conformance-test-ci [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+#   - CLOUD_PROVIDER 
+#     cloud provider (az, os, gcp, aws)
+#   - SECRET_BINDING
+#     name of secret binding
+#   - SHOOT_DOMAIN
+#     name of shoot domain, e.g. example.com
+#
+# -----------------------------------------------------------------------------
+
+TRUE=0
```
Comment (msohn) [inline .ci/run-conformance-test-ci]: declare these variables to be constant ?

readonly TRUE=0

File: .ci/run-conformance-test-ci [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+# Sript parameter
+VALID_CLOUD_PROVIDERS="aws gcp os az"
+
+# -----------------------------------------------------------------------------
+# Helper functions
+
+# -----------------------------------------------------------------------------
+# help
+function help() {
+  echo "run-conformance-test-ci <garden-kubeconfig> <garden-namespace> <cloud-provider> <secret-binding> <shoot-domain>"
```
Comment (msohn) [inline .ci/run-conformance-test-ci]: use a document and add some more text to improve readability ?
```bash
cat << EOF
Usage: 
run-conformance-test-ci <garden-kubeconfig> <garden-namespace> <cloud-provider> <secret-binding> <shoot-domain>

Parameters:
  <garden-kubeconfig>: path to kubeconfig of Garden cluster
  <garden-namespace>: string, e.g. garden-mytest. Naming convention: garden-<project>
  <cloud-provider>: string, one of (${VALID_CLOUD_PROVIDERS})
  <secret-binding>: string, name of the secret binding to be used in the shoot manifest
  <shoot-domain>: string, name of the domain to be used in the shoot manifest, e.g. example.com

Description: Runs kubernetes conformance tests

Example:
   run-conformance-test-ci ./foo/kubeconfig garden-mytest aws aws-secret-binding example.com
EOF
```

File: .ci/run-conformance-test-ci [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+  if ! create_shoot ; then
+    echo "ERROR: Shoot could not be created."
+    exit 1
+  fi
+
+  # Deploy and run conformance test
+  echo "INFO: Deploy the conformance test case"
+  curl -sL https://raw.githubusercontent.com/cncf/k8s-conformance/master/sonobuoy-conformance.yaml | kubectl --kubeconfig="$SHOOT_KUBECONFIG" apply -f - > /dev/null
+
+  echo  "INFO: Wait until the pod sonobuoy/sonobuoy cecomes available ... or MAX_RETRY_TIME="$MAX_RETRY_TIME "s exceeded"
```
Comment (msohn) [inline .ci/run-conformance-test-ci]: s/cecomes/becomes/

File: .ci/run-conformance-test-ci [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+#   MAX_RETRY_TIME
+function run_conformance_tests() {
+
+  # Create shoot
+  if ! create_shoot ; then
+    echo "ERROR: Shoot could not be created."
+    exit 1
+  fi
+
+  # Deploy and run conformance test
```
Comment (msohn) [inline .ci/run-conformance-test-ci]: extract this into functions deploy_test and run_test ?

File: .ci/run-conformance-test-ci [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+    fi
+  done
+
+  if [ "$success" -eq "$FALSE" ]; then
+    echo "ERROR: Conformance tests didn't finish in  ${MAX_TEST_TIME} seconds!"
+    delete_shoot
+    exit 1
+  fi
+
+  # Copy the test result to local drive
```
Comment (msohn) [inline .ci/run-conformance-test-ci]: extract into function copy_test_results

File: .ci/run-conformance-test-ci [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+
+  if [[ "$RESULT_FILENAME" = *"tar.gz"* ]]; then
+    kubectl --kubeconfig="$SHOOT_KUBECONFIG" cp sonobuoy/sonobuoy:/tmp/sonobuoy/"$RESULT_FILENAME" "$RESULT_DIR" 2> /dev/null
+  else
+    echo "ERROR: Conformance test failed: no expected *.tar.gz archive in /tmp/sonobuoy generated" >&2
+    delete_shoot
+    exit 1
+  fi
+
+  # Examine e2e.log
```
Comment (msohn) [inline .ci/run-conformance-test-ci]: extract into function check_test_result ?

File: .ci/run-conformance-test-ci [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+mkdir -p "$RESULT_DIR"
+
+SHOOT_KUBECONFIG="${RESULT_DIR}/${CLOUD_PROVIDER}_kubeconfig.yaml"
+
+# Echo conformance test result
+if ! run_conformance_tests ; then
+  echo "INFO: Conformance test SUCCESSFULLY finished"
+  exit 0
+else
+  echo "INFO: Conformance test NOT successfully finished"
```
Comment (msohn) [inline .ci/run-conformance-test-ci]: how about this: Conformance test FAILED

File: .ci/shoot-utils [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+    echo "ERROR: shoot creation failed (mako-render and/or kubectl)"
+    return "$success"
+  fi
+
+  echo  "INFO: Wait until the shoot becomes available ... or MAX_RETRY_TIME=${MAX_RETRY_TIME}s exceeded"
+  retry_stop=$(($(date +%s) + $MAX_RETRY_TIME))
+  wait_time=0
+  while [[ $(date +%s) -lt $retry_stop ]]; do
+    sleep 15
+    # Check status.lastOperation.status
```
Comment (msohn) [inline .ci/shoot-utils]: extract into function check_last_operation_status ?

File: .ci/shoot-utils [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+#     name of shoot domain, e.g. example.com
+#   - GARDEN_KUBECONFIG
+#     Path to the Gardener Kubeconfig
+#   - MAX_RETRY_TIME
+#     integer: time in seconds to wait for shoot creation/deletion
+# -----------------------------------------------------------------------------
+
+
+# -----------------------------------------------------------------------------
+# create_shoot
```
Comment (msohn) [inline .ci/shoot-utils]: why is the function name repeated here ?

File: .ci/run-conformance-test-ci [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+
+# Read environment variables
+CONFORMANCE_TEST_PATH=$(readlink -f "${CONFORMANCE_TEST_PATH}")
+GARDENER_PATH=$(readlink -f "${GARDENER_PATH}")
+
+# Sript parameter
+VALID_CLOUD_PROVIDERS="aws gcp os az"
+
+# -----------------------------------------------------------------------------
+# Helper functions
```
Comment (msohn) [inline .ci/run-conformance-test-ci]: this comment doesn't seem to add value

File: .ci/run-conformance-test-ci [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+  echo "  <garden-namespace>: string, e.g. garden-mytest. Naming convention: garden-<project>"
+  echo "  <cloud-provider>: string, one of (${VALID_CLOUD_PROVIDERS})"
+  echo "  <secret-binding>: string, name of the secret binding to be used in the shoot manifest"
+  echo "  <shoot-domain>: string, name of the domain to be used in the shoot manifest, e.g. example.com"
+  echo "Example:"
+  echo "   run-conformance-test-ci ./foo/kubeconfig garden-mytest aws aws-secret-binding example.com"
+}
+
+# -----------------------------------------------------------------------------
+# trap_handler
```
Comment (msohn) [inline .ci/run-conformance-test-ci]: unnecessary comment

File: .ci/run-conformance-test-ci [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+       delete_shoot
+     fi
+    else
+     echo "ERROR: kubectl failed in trap_handler. Call delete_shoot anyway in case of a temporary issue."
+     delete_shoot
+   fi  
+}
+
+# -----------------------------------------------------------------------------
+# run_conformance_tests
```
Comment (msohn) [inline .ci/run-conformance-test-ci]: unnecessary repetition ?

**Review by msohn — CHANGES_REQUESTED** [review]

**Comment by rfranzke** [comment]
Apart from @msohn's comments I don't have any more feedback for the code itself yet (as discussed, we plan to replace the test scripts with Golang anyway at some point in time).
Regarding the integration test, @dkistner does not plan to use any of your "new" functionality but he will just replenish the old script (as also this test script will be rewritten in Go instead of mixed bash/python).

Therefore, you can move the functions which you put in a separate file (`shoot-utils`) in the original test script and do not need to source any more. Moreover, let's rename the whole script from `run-conformance-test-ci` to `conformance-test` or `k8s-conformance-test`, please.

Thank you a lot, @rolpat!!

**Comment by rfranzke** [comment]
Could you please squash the commits?

File: .ci/k8s-conformance-test [hunk 13]
```diff
@@ -0,0 +1,368 @@
...
+  <garden-kubeconfig>: path to kubeconfig of Garden cluster
+  <garden-namespace>: string, e.g. garden-mytest. Naming convention: garden-<project>
+  <cloud-provider>: string, one of (${VALID_CLOUD_PROVIDERS})
+  <secret-binding>: string, name of the secret binding to be used in the shoot manifest
+  <shoot-domain>: string, name of the domain to be used in the shoot manifest, e.g. example.com
+
+Description: Runs kubernetes conformance tests
+
+Example:
+   run-conformance-test-ci ./foo/kubeconfig garden-mytest aws aws-secret-binding example.com
```
Comment (rfranzke) [inline .ci/k8s-conformance-test]: Rename `run-conformance-test-ci` into `k8s-conformance-test`.

File: .ci/k8s-conformance-test [hunk 14]
```diff
@@ -0,0 +1,368 @@
...
+GARDENER_PATH=$(readlink -f "${GARDENER_PATH}")
+
+# Sript parameter
+VALID_CLOUD_PROVIDERS="aws gcp os az"
+
+# -----------------------------------------------------------------------------
+function help() {
+cat << EOF
+Usage: 
+run-conformance-test-ci <garden-kubeconfig> <garden-namespace> <cloud-provider> <secret-binding> <shoot-domain>
```
Comment (rfranzke) [inline .ci/k8s-conformance-test]: Rename `run-conformance-test-ci` into `k8s-conformance-test`.

**Review by rfranzke — APPROVED** [review]

**Pull request merged**